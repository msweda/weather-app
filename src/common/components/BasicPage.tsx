import {
  createStyles,
  Theme,
  WithStyles,
  withStyles,
} from '@material-ui/core/styles'
import * as React from 'react'

const BasicPage: React.FC<Props> = ({ classes, children }) => {
  return (
    <div className={classes.basicPage}>
      <div className={classes.leftPageMargin} />
      <div className={classes.pageContent}>{children}</div>
      <div className={classes.rightPageMargin} />
    </div>
  )
}

const styles = (theme: Theme) =>
  createStyles({
    basicPage: {
      flex: 1,
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'center',
      alignItems: 'stretch',
      backgroundColor: theme.palette.background.paper,
    },
    leftPageMargin: {
      flex: 1,
    },
    pageContent: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'stretch',
      width: 800,
      [theme.breakpoints.down('sm')]: {
        width: 600,
      },
      [theme.breakpoints.down('xs')]: {
        width: 300,
      },
      padding: 20,
    },
    rightPageMargin: {
      flex: 1,
    },
  })

interface Props extends WithStyles<typeof styles> {}

export default withStyles(styles)(BasicPage)
