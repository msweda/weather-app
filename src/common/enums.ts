export enum RootRoute {
  Home = '/',
  Weather = '/weather',
  Forecast = '/forecast',
}

export enum AsyncStatus {
  Idle = 'IDLE',
  InProgress = 'IN_PROGRESS',
  Succeeded = 'SUCCEEDED',
  Failed = 'FAILED',
}
