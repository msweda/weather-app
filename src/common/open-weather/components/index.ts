export { default as UnitSelect } from './UnitSelect'
export { default as Weather } from './Weather'
