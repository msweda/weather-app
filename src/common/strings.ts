export const appHeading = 'Matthew Sweda\'s Weather App';
export const phoneNumber = '(248)-770-7487';
export const email = 'msweda.developer@gmail.com';
export const location = 'Detroit, MI';
export const copyright = (timestamp: number) => `Â© Matthew Sweda ${new Date(timestamp).getFullYear()}`;
export const emailLink = `mailto:${email}`;
export const linkedInLink = 'https://www.linkedin.com/in/matthew-sweda-118570ba/';
export const instagramLink = 'https://www.instagram.com/mattsweda/';