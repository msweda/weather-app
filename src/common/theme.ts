// create palette here: https://material.io/color/

import { createMuiTheme } from '@material-ui/core/styles'

export default createMuiTheme({
  typography: {
    useNextVariants: true,
  },
  palette: {
    primary: {
      main: '#2d4671',
    },
    secondary: {
      main: '#6ba099',
    },
    background: {
      default: '#2d4671',
      paper: '#eee',
    },
    text: {
      primary: '#fff',
      secondary: '#000',
    },
  },
  overrides: {
    MuiCssBaseline: {
      '@global': {
        body: {
          display: 'flex',
          minHeight: '100vh',
        },
        '#app': {
          flex: 1,
          display: 'flex',
          flexDirection: 'column',
        },
      },
    },
  },
})
