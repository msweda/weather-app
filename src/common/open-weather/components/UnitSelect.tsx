import { UnitType } from '@/common/open-weather/enums'
import { AllStores } from '@/src/stores'
import FormControl from '@material-ui/core/FormControl'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import Select from '@material-ui/core/Select'
import {
  createStyles,
  Theme,
  WithStyles,
  withStyles,
} from '@material-ui/core/styles'
import { inject, observer } from 'mobx-react'
import * as React from 'react'
import OpenWeatherStore from '../store'

@inject((allStores: AllStores) => ({
  openWeatherStore: allStores.openWeatherStore,
}))
@observer
class UnitSelect extends React.Component<Props> {
  get injected() {
    return this.props as InjectedProps
  }

  render() {
    const { classes } = this.props
    return (
      <FormControl className={classes.selectFormControl}>
        <InputLabel htmlFor="weather-units-input">Units</InputLabel>
        <Select
          className={classes.select}
          value={this.injected.openWeatherStore.unitType}
          onChange={this.changeUnitType}
          inputProps={{ id: 'weather-units-input' }}
        >
          <MenuItem className={classes.menuItem} value={UnitType.Scientific}>
            Kelvin
          </MenuItem>
          <MenuItem className={classes.menuItem} value={UnitType.Metric}>
            Celsius
          </MenuItem>
          <MenuItem className={classes.menuItem} value={UnitType.Imperial}>
            Fahrenheit
          </MenuItem>
        </Select>
      </FormControl>
    )
  }

  private changeUnitType = (event: React.ChangeEvent<HTMLSelectElement>) => {
    this.injected.openWeatherStore.setUnitType(event.target.value as UnitType)
    this.injected.openWeatherStore.clearWeather()
    this.injected.openWeatherStore.clearForecast()
  }
}

const styles = (theme: Theme) =>
  createStyles({
    selectFormControl: {
      marginBottom: 20,
    },
    select: {
      color: theme.palette.primary.main,
    },
    menuItem: {
      color: theme.palette.primary.main,
    },
  })

interface Props extends WithStyles<typeof styles> {}

interface InjectedProps extends Props {
  openWeatherStore: OpenWeatherStore
}

export default withStyles(styles)(UnitSelect)
