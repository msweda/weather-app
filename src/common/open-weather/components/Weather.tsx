import { createStyles, WithStyles, withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import * as React from 'react'
import { Weather as WeatherInterface } from '../interfaces'
import { weatherIconUrl } from '../strings'

class Weather extends React.Component<Props> {
  render() {
    const { classes, weather } = this.props
    return (
      <div className={classes.weatherIcons}>
        {weather.map((element, index) => (
          <div
            key={`weather-array-element-${index}`}
            className={classes.weatherArrayElement}
          >
            <img src={weatherIconUrl(element.icon)} width="30" height="30" />
            <Typography color="primary">{element.main}</Typography>
          </div>
        ))}
      </div>
    )
  }
}

const styles = () =>
  createStyles({
    weatherIcons: {
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'center',
      alignItems: 'center',
    },
    weatherArrayElement: {
      margin: 10,
    },
  })

interface Props extends WithStyles<typeof styles> {
  weather: WeatherInterface
}

export default withStyles(styles)(Weather)
