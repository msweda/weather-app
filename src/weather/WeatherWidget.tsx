import { UnitSelect, Weather } from '@/common/open-weather/components'
import OpenWeatherStore from '@/common/open-weather/store'
import { noGeolocation, weatherFailed } from '@/common/open-weather/strings'
import { isGeolocationAvailable } from '@/common/utils'
import { AllStores } from '@/src/stores'
import Button from '@material-ui/core/Button'
import CircularProgress from '@material-ui/core/CircularProgress'
import { createStyles, WithStyles, withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import { inject, observer } from 'mobx-react'
import * as React from 'react'

@inject((allStores: AllStores) => ({
  openWeatherStore: allStores.openWeatherStore,
}))
@observer
class WeatherWidget extends React.Component<Props> {
  get injected() {
    return this.props as InjectedProps
  }

  render() {
    if (!isGeolocationAvailable()) {
      return <Typography>{noGeolocation}</Typography>
    }

    const { classes } = this.props

    return (
      <React.Fragment>
        <div className={classes.userInput}>
          <UnitSelect />
          {this.renderGetWeatherButton()}
        </div>
        {this.renderWeatherBlock()}
      </React.Fragment>
    )
  }

  private renderGetWeatherButton = () => {
    if (this.injected.openWeatherStore.weatherInProgress) {
      return <CircularProgress />
    }
    return (
      <Button variant="contained" color="primary" onClick={this.getWeather}>
        Get Weather
      </Button>
    )
  }

  private renderWeatherBlock = () => {
    if (
      this.injected.openWeatherStore.weatherSucceeded &&
      this.injected.openWeatherStore.weatherData !== undefined
    ) {
      const { classes } = this.props
      return (
        <div className={classes.weatherBlock}>
          <Typography variant="h6" color="textSecondary">
            The current temperature in your area is:
          </Typography>
          <Typography variant="h1" color="primary">
            {this.injected.openWeatherStore.weatherData.main.temp}
          </Typography>
          <Typography variant="h6" color="textSecondary">
            Conditions:
          </Typography>
          <Weather
            weather={this.injected.openWeatherStore.weatherData.weather}
          />
        </div>
      )
    }

    if (this.injected.openWeatherStore.weatherFailed) {
      return <Typography>{weatherFailed}</Typography>
    }

    return null
  }

  private getWeather = () => {
    this.injected.openWeatherStore.getWeather()
  }
}

const styles = () =>
  createStyles({
    userInput: {
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'center',
      alignItems: 'center',
      marginBottom: 20,
      '& > *': {
        margin: 10,
      },
    },
    weatherBlock: {
      textAlign: 'center',
    },
  })

interface Props extends WithStyles<typeof styles> {}

interface InjectedProps extends Props {
  openWeatherStore: OpenWeatherStore
}

export default withStyles(styles)(WeatherWidget)
