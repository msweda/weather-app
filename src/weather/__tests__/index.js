import { BasicPage } from '@/common/components'
import { createShallow } from '@material-ui/core/test-utils'
import * as React from 'react'
import Weather from '../index'
import WeatherWidget from '../WeatherWidget'

describe('<Weather />', () => {
  let shallow
  let wrapper

  beforeEach(() => {
    shallow = createShallow({ untilSelector: Weather })
    wrapper = shallow(<Weather />)
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render a <BasicPage> that contains a <WeatherWidget>', () => {
    const basicPage = wrapper.find(BasicPage)
    expect(basicPage.exists()).toBeTruthy()
    expect(basicPage.contains(<WeatherWidget />)).toBeTruthy()
  })
})
