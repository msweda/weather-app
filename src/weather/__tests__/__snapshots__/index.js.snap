// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Weather /> should match snapshot 1`] = `
<div>
  <WithStyles(BasicPage)>
    <WithStyles(inject-WeatherWidget) />
  </WithStyles(BasicPage)>
</div>
`;
